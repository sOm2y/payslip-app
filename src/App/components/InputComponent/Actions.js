//@flow

import React from 'react';
import Button from 'material-ui/Button';

class Actions extends React.PureComponent<{ onClickOk: Function, onClickReset: Function }> {
  render() {
    const { onClickOk, onClickReset } = this.props;

    return (
      <div
        style={{
          display: 'flex',
          justifyContent: 'space-between'
        }}
      >
        <Button className="btn_reset" onClick={onClickReset}>
          Reset
        </Button>

        <Button className="btn_ok" color="primary" raised onClick={onClickOk}>
          Ok
        </Button>
      </div>
    );
  }
}

export default Actions;
