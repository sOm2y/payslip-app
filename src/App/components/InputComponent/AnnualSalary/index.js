//@flow

export { default } from './AnnualSalary';
