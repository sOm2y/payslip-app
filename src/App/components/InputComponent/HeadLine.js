//@flow

import React from 'react';
import Typography from 'material-ui/Typography';

class HeadLine extends React.PureComponent<{}> {
  render() {
    return (
      <div
        style={{
          height: 50
        }}
      >
        <Typography type="headline" id="title">
          Please input your detail
        </Typography>
      </div>
    );
  }
}

export default HeadLine;
