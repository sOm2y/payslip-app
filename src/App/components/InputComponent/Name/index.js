//@flow

export { default } from './Name';
