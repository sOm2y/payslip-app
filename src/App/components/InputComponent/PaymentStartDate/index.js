//@flow

export { default } from './PaymentStartDate';
