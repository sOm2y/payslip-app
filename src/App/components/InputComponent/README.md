# input component

- should have a number of input text areas / buttons to accept user input
- should do data validation based on input
- should response input errors when something goes wrong

fields:
first name -> input , string
last name -> input , string
annual salary -> input , positive, money value
super rate -> input, percentage, 0-50 inclusive
payment start date -> date