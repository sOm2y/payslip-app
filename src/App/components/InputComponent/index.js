//@flow

export { default } from './InputComponent';
