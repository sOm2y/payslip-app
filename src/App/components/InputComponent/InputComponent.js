//@flow

/**
 * the input component
 * 
 */

import React from 'react';
import Grid from 'material-ui/Grid';

import Name from './Name';
import AnnualSalary from './AnnualSalary';
import SuperRate from './SuperRate';
import PaymentStartDate from './PaymentStartDate';
import HeadLine from './HeadLine';
import Actions from './Actions';
import { doMapState } from './shared';

import Spacer from '../shared/Spacer';
import Container from '../shared/Container';

type Props = {
  onSubmit: PayslipInput => PayslipCalculationResult,
  onReset: () => mixed
};

const initialState: InputComponentStates = {
  firstname: '',
  lastname: '',
  annualSalary: '',
  superRate: '',
  paymentStartDate: '',
  failReasons: {}
};

const renderInputArea = (renderInput: Function) => (
  <Grid container id="input-area">
    <Grid item xs={12}>
      {renderInput(Name)('firstname')}
    </Grid>
    <Grid item xs={12}>
      {renderInput(Name)('lastname')}
    </Grid>
    <Grid item xs={12} md={8}>
      {renderInput(AnnualSalary)('annualSalary')}
    </Grid>
    <Grid item xs={12} md={4}>
      {renderInput(SuperRate)('superRate')}
    </Grid>

    <Grid item xs={12}>
      {renderInput(PaymentStartDate)('paymentStartDate')}
    </Grid>
  </Grid>
);

class InputComponent extends React.PureComponent<Props, InputComponentStates> {
  state = initialState;

  _updateField = (field: FieldTypes) => (value: string) => {
    this.setState({ [field]: value, failReasons: { ...this.state.failReasons, [field]: '' } });
  };

  _renderInput = (Comp: Class<React.Component<InputProps>>) => (field: FieldTypes) => (
    <Comp
      type={field}
      key={field}
      value={this.state[field]}
      onChange={this._updateField(field)}
      failReasons={this.state.failReasons[field]}
    />
  );

  _handleSubmit = () => {
    const result = this.props.onSubmit(doMapState(this.state));
    if (result.isInvalid) {
      this.setState({ failReasons: result.reasons });
    }
  };
  render() {
    const { onReset } = this.props;

    return (
      <Container>
        <HeadLine />
        {renderInputArea(this._renderInput)}
        <Spacer height={72} />
        <Actions
          onClickReset={() => {
            this.setState(initialState, () => {
              onReset();
            });
          }}
          onClickOk={this._handleSubmit}
        />
      </Container>
    );
  }
}

export default InputComponent;
