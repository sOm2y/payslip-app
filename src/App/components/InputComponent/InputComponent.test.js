//@flow

import React from 'react';
import InputComponent from './InputComponent';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import '../../setupTest'; // enzyme setup

describe('Actions', () => {
  it('should renders without crashing even for shallow renderer', () => {
    shallow(<InputComponent onSubmit={jest.fn()} />);
  });

  it('should have ok action', () => {
    const onSubmitStub = jest.fn();
    const state = {
      firstname: 'foo',
      lastname: 'bar',
      annualSalary: '60050',
      superRate: '0.09',
      paymentStartDate: '3',
      failReasons: {}
    };
    onSubmitStub.mockReturnValue({
      isInvalid: false
    });
    const sut = mount(<InputComponent onSubmit={onSubmitStub} onReset={jest.fn()} />);

    sut.setState(state);
    sut
      .find('.btn_ok')
      .first()
      .simulate('click');

    expect(onSubmitStub).toHaveBeenCalledTimes(1);
  });

  it('should have reset action', () => {
    const onResetStub = jest.fn();
    const state = {
      firstname: 'foo',
      lastname: 'bar',
      annualSalary: '60050',
      superRate: '0.09',
      paymentStartDate: '3',
      failReasons: {}
    };

    const sut = mount(<InputComponent onSubmit={jest.fn()} onReset={onResetStub} />);

    sut.setState(state);
    sut
      .find('.btn_reset')
      .first()
      .simulate('click');

    expect(onResetStub).toHaveBeenCalledTimes(1);

    expect(sut.state()).toEqual({
      firstname: '',
      lastname: '',
      annualSalary: '',
      superRate: '',
      paymentStartDate: '',
      failReasons: {}
    });
  });

  it('should be able to accept input', () => {
    const sut = mount(<InputComponent onSubmit={jest.fn()} onReset={jest.fn()} />);

    // the first input field should be first name, because it is on the top
    sut
      .find('input')
      .first()
      .simulate('change', { target: { value: 'foo', focus: () => {} } });

    expect(sut.state().firstname).toBe('foo');
  });

  it('should response to failure', () => {
    const onSubmitStub = jest.fn();

    const failReason = { firstname: 'foo' };

    onSubmitStub.mockReturnValue({
      isInvalid: true,
      reasons: failReason
    });

    const sut = mount(<InputComponent onSubmit={onSubmitStub} onReset={jest.fn()} />);

    sut
      .find('.btn_ok')
      .first()
      .simulate('click');
    expect(sut.state().failReasons).toBe(failReason);
  });

  it('should match snapshot', () => {
    const tree = renderer.create(<InputComponent onSubmit={jest.fn()} onReset={jest.fn()} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
