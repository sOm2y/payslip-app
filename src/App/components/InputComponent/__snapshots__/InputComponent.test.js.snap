// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions should match snapshot 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-shadow2-6 MuiPaper-rounded-3 Container-container-1"
>
  <div
    style={
      Object {
        "height": 50,
      }
    }
  >
    <h1
      className="MuiTypography-root-29 MuiTypography-headline-34"
      id="title"
    >
      Please input your detail
    </h1>
  </div>
  <div
    className="MuiGrid-typeContainer-53 MuiGrid-spacing-xs-16-73"
    id="input-area"
  >
    <div
      className="MuiGrid-typeItem-54 MuiGrid-grid-xs-12-88"
    >
      <div
        className="MuiFormControl-root-141"
        onBlur={[Function]}
        onFocus={[Function]}
        required={false}
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <label
          className="MuiFormLabel-root-151 MuiInputLabel-root-145 MuiInputLabel-formControl-146 MuiInputLabel-animated-149"
          data-shrink={false}
        >
          First Name
        </label>
        <div
          className="MuiInput-root-155 MuiInput-formControl-156 MuiInput-inkbar-157 MuiInput-underline-163"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <input
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-159 MuiInput-inputSingleline-166"
            defaultValue={undefined}
            disabled={false}
            id={undefined}
            name={undefined}
            onChange={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-typeItem-54 MuiGrid-grid-xs-12-88"
    >
      <div
        className="MuiFormControl-root-141"
        onBlur={[Function]}
        onFocus={[Function]}
        required={false}
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <label
          className="MuiFormLabel-root-151 MuiInputLabel-root-145 MuiInputLabel-formControl-146 MuiInputLabel-animated-149"
          data-shrink={false}
        >
          Last Name
        </label>
        <div
          className="MuiInput-root-155 MuiInput-formControl-156 MuiInput-inkbar-157 MuiInput-underline-163"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <input
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-159 MuiInput-inputSingleline-166"
            defaultValue={undefined}
            disabled={false}
            id={undefined}
            name={undefined}
            onChange={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-typeItem-54 MuiGrid-grid-xs-12-88 MuiGrid-grid-md-8-110"
    >
      <div
        className="MuiFormControl-root-141"
        onBlur={[Function]}
        onFocus={[Function]}
        required={false}
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <label
          className="MuiFormLabel-root-151 MuiInputLabel-root-145 MuiInputLabel-formControl-146 MuiInputLabel-animated-149 MuiInputLabel-shrink-148"
          data-shrink={true}
        >
          Annual Salary
        </label>
        <div
          className="MuiInput-root-155 MuiInput-formControl-156 MuiInput-inkbar-157 MuiInput-underline-163"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <div
            className="MuiInputAdornment-root-170 MuiInputAdornment-positionStart-171"
          >
            <p
              className="MuiTypography-root-29 MuiTypography-body1-38 MuiTypography-colorSecondary-50"
            >
              $
            </p>
          </div>
          <input
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-159 MuiInput-inputSingleline-166"
            defaultValue={undefined}
            disabled={false}
            id={undefined}
            name={undefined}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={undefined}
            onMouseUp={[Function]}
            placeholder={undefined}
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-typeItem-54 MuiGrid-grid-xs-12-88 MuiGrid-grid-md-4-106"
    >
      <div
        className="MuiFormControl-root-141"
        onBlur={[Function]}
        onFocus={[Function]}
        required={false}
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <label
          className="MuiFormLabel-root-151 MuiInputLabel-root-145 MuiInputLabel-formControl-146 MuiInputLabel-animated-149"
          data-shrink={false}
        >
          Super Rate
        </label>
        <div
          className="MuiInput-root-155 MuiInput-formControl-156 MuiInput-inkbar-157 MuiInput-underline-163"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <input
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-159 MuiInput-inputSingleline-166"
            defaultValue={undefined}
            disabled={false}
            id={undefined}
            name={undefined}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={undefined}
            onMouseUp={[Function]}
            placeholder={undefined}
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type="text"
            value=""
          />
          <div
            className="MuiInputAdornment-root-170 MuiInputAdornment-positionEnd-172"
          >
            <p
              className="MuiTypography-root-29 MuiTypography-body1-38 MuiTypography-colorSecondary-50"
            >
              %
            </p>
          </div>
        </div>
        <p
          className="MuiFormHelperText-root-173"
        >
          between 0 and 50
        </p>
      </div>
    </div>
    <div
      className="MuiGrid-typeItem-54 MuiGrid-grid-xs-12-88"
    >
      <div
        className="MuiFormControl-root-141"
        onBlur={[Function]}
        onFocus={[Function]}
        required={false}
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <div
          className="MuiFormControl-root-141 MuiFormControl-marginNormal-142"
          onBlur={[Function]}
          onFocus={[Function]}
          required={false}
        >
          <label
            className="MuiFormLabel-root-151 MuiInputLabel-root-145 MuiInputLabel-formControl-146 MuiInputLabel-animated-149"
            data-shrink={false}
            htmlFor={undefined}
          >
            Payment Start Date
          </label>
          <div
            className="MuiInput-root-155 MuiInput-formControl-156 MuiInput-inkbar-157 MuiInput-underline-163"
            onBlur={[Function]}
            onFocus={[Function]}
          >
            <div
              className="MuiSelect-root-177"
            >
              <select
                autoComplete={undefined}
                autoFocus={undefined}
                className="MuiSelect-select-178 MuiInput-input-159 MuiInput-inputSingleline-166"
                defaultValue={undefined}
                disabled={false}
                id={undefined}
                name={undefined}
                onBlur={undefined}
                onChange={[Function]}
                onFocus={undefined}
                onKeyDown={undefined}
                onKeyUp={undefined}
                placeholder={undefined}
                readOnly={undefined}
                required={undefined}
                rows={undefined}
                type="text"
                value=""
              >
                <option
                  value={-1}
                >
                   
                </option>
                <option
                  value="10"
                >
                  1st November - 30th November, 2017
                </option>
                <option
                  value="11"
                >
                  1st December - 31st December, 2017
                </option>
                <option
                  value="0"
                >
                  1st January - 31st January, 2018
                </option>
                <option
                  value="1"
                >
                  1st February - 28th February, 2018
                </option>
                <option
                  value="2"
                >
                  1st March - 31st March, 2018
                </option>
                <option
                  value="3"
                >
                  1st April - 30th April, 2018
                </option>
                <option
                  value="4"
                >
                  1st May - 31st May, 2018
                </option>
                <option
                  value="5"
                >
                  1st June - 30th June, 2018
                </option>
                <option
                  value="6"
                >
                  1st July - 31st July, 2018
                </option>
                <option
                  value="7"
                >
                  1st August - 31st August, 2018
                </option>
                <option
                  value="8"
                >
                  1st September - 30th September, 2018
                </option>
                <option
                  value="9"
                >
                  1st October - 31st October, 2018
                </option>
              </select>
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-182 MuiSelect-icon-181"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "height": 72,
        "width": undefined,
      }
    }
  />
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <button
      className="MuiButtonBase-root-197 MuiButton-root-183 btn_reset"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label-185"
      >
        Reset
      </span>
      <span
        className="MuiTouchRipple-root-199"
      />
    </button>
    <button
      className="MuiButtonBase-root-197 MuiButton-root-183 MuiButton-raised-190 MuiButton-raisedPrimary-192 btn_ok"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label-185"
      >
        Ok
      </span>
      <span
        className="MuiTouchRipple-root-199"
      />
    </button>
  </div>
</div>
`;
